File: 150.evaluate-reverse-polish-notation.py
#
# @lc app=leetcode id=150 lang=python3
#
# [150] Evaluate Reverse Polish Notation
#

# Time:  O(n)
# Space: O(n)

from typing import List


# @lc code=start
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        def calc(n1, n2, op):
            if op == "+":
                return n1 + n2
            elif op == "-":
                return n1 - n2
            elif op == "*":
                return n1 * n2
            else:
                return int(n1 / n2)

        stack, ops = [], ["+", "-", "*", "/"]
        for token in tokens:
            if token not in ops:
                stack.append(int(token))
            else:
                n2, n1 = stack.pop(), stack.pop()
                stack.append(calc(n1, n2, token))
        return stack[0]


# @lc code=end

File: 150.evaluate-reverse-polish-notation.js
/*
 * @lc app=leetcode id=150 lang=javascript
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function (tokens) {
  const calc = (n1, n2, op) => {
    if (op === "+") {
      return n1 + n2;
    } else if (op === "-") {
      return n1 - n2;
    } else if (op === "*") {
      return n1 * n2;
    } else if (op === "/") {
      return Math.trunc(n1 / n2);
    }
  };
  const stk = [];
  const ops = "+-*/";
  for (const token of tokens) {
    if (!ops.includes(token)) {
      stk.push(parseInt(token));
    } else {
      const n2 = stk.pop();
      const n1 = stk.pop();
      stk.push(calc(n1, n2, token));
    }
  }
  return stk[0];
};
// @lc code=end
